Delete runs in daterange in batches

Roemer opened the pull request:
This PR changes the behavior of the deletion of runs in a specified date range.
Previously, it used a very slow `aggregate` to find the `instances` for the `runs` that should be deleted and tried to delete them all at once. This timeouted at various ends in larger setups.
With this change, it will now process the deletion in batches:
1. Get a batch of `runIds`
2. Delete all associated `instances` for those `runIds`
3. Delete the `runs` for those `runIds`
4. Repeat until no more `runs` are found for the defined date-range
5. 
To make it even faster, a new index for `runId` is added to the `instance` collection.

agoldis commented on packages/api/src/datasources/runs.ts:
```diff
@@ -136,23 +136,37 @@ export class RunsAPI extends DataSource {
   }
 
-  async deleteRunsInDateRange(startDate: Date, endDate: Date) {
+  async deleteRunsInDateRange(
+    startDate: Date,
+    endDate: Date,
+    limit: number = 0
+  ) {
+    const getResult = await this.getRunsInDateRange(startDate, endDate, limit);
+    return this.deleteRunsByIds(getResult.runIds);
+  }
```
This one is not used, please delete

agoldis reviewed (approved):
Great job @Roemer! Thanks a lot for fixing that code I am happy to see that people use Sorry Cypress to run as many test that we are facing performance issues 😬

Just a few minor comments:

- batched iteration on results is great and simple! I think the same result could be achieved with native mongoDB [cursor](https://docs.mongodb.com/drivers/node/fundamentals/crud/read-operations/cursor/#asynchronous-iteration), but your implementation is totally find!
- if we are really looking to optimize it, we can limit the fields we get from `getRunsInDateRange` by explicitly [fetching **only**](https://docs.mongodb.com/drivers/node/fundamentals/crud/read-operations/project/#specify-which-fields-to-return) `runId`s.   If I am not mistaken, mongoDB won't even read data from the disk and send the results back straight from the `runId` index on runs collection.

agoldis commented on packages/api/src/datasources/instances.ts:
```diff
@@ -89,32 +89,6 @@ export class InstancesAPI extends DataSource {
-
-    const runIds = response.map((x) => x.run[0].runId) as string[];
-    return await this.deleteInstancesByRunIds(runIds);
-  }
```
👍🏻

agoldis commented on packages/api/src/resolvers/index.ts:
```diff
@@ -160,15 +160,36 @@ export const resolvers = {
+          break;
+        }
+        // Delete the runs
+        const runsDeleteResponse = await resolvers.Mutation.deleteRuns(
```
Nice reuse of mutation 👍🏻

agoldis commented:
P.S. I've merged #427 into master, don't be scared

Roemer commented:
I rebased it back to master instead of merged and removed the unneeded method and added a projection when getting the runs.
Good point about the cursor. I had a look at it and it would not really fit well into the current architecture. We would need to add the cursor to datasources/run.ts and actually return it so that we can handle the deletion of the instances, runs and runtimeouts in the mutation itself. But then we would have a datasource related object (the cursor) in the mutation itself which seems wrong. Or we would pass a function to the run datasource which would handle that all, but that also seems not so nice. So I think for a first version, we can go with this.
Testing is still pending on my side. I will probably need to do that tomorrow.

Roemer commented:
I now tested it and I could delete 2 months worth of data in a few seconds ;) So from my side, this PR can be merged.
